### PR TITLE
Introduced abstract ``NurbsBase``

### LINKED ISSUE #334: Add ``NurbsBase`` abstract class.

### PR DESCRIPTION
**Pull request by sonomirco**
### What type of PR is this? (check all applicable)

- [X] ♻️ Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

### Description

This PR introduces the abstract class ``NurbsBase`` that will be used to implement ``Polycurve``.
``NurbsBase`` could be inherited from all the other curve classes overriding the methods where trigonometry is handy.

### Related Tickets & Documents

This PR closes #334 

### Added tests?

- [ ] 👍 yes
- [X] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📓 docs
- [X] 🙅 no documentation needed

### CODE REVIEW

File: src/GShark/Geometry/NurbsBase.cs [hunk 1]
```diff
@@ -0,0 +1,705 @@
...
+        /// <summary>
+        /// Calculates the 3D plane at the given parameter.
+        /// Defined as the Frenet frame, is constructed from the velocity and the acceleration of the curve.
+        /// https://janakiev.com/blog/framing-parametric-curves/
+        /// </summary>
+        /// <param name="t">Evaluation parameter. Parameter should be between 0.0 and 1.0.</param>
+        /// <returns>The perpendicular frame.</returns>
+        public Plane FrameAt(double t)
+        {
+            if (t <= 0.0)
```
Comment (cesarecaoduro) [inline src/GShark/Geometry/NurbsBase.cs:278]: Maybe the name should be perpendicular frame like in Rhino?

File: src/GShark/Geometry/NurbsBase.cs [hunk 2]
```diff
@@ -0,0 +1,705 @@
...
+
+            return curves;
+        }
+
+        /// <summary>
+        /// Extract sub-curve defined by domain.
+        /// </summary>
+        /// <param name="domain">Domain of sub-curve</param>
+        /// <returns>NurbsBase.</returns>
+        public NurbsBase SubCurve(Interval domain)
```
Comment (cesarecaoduro) [inline src/GShark/Geometry/NurbsBase.cs:587]: In the future would also be useful to have multiple splits with a list of parameters or a list of lengths

File: src/GShark/Optimization/CurvesIntersectionObjectives.cs [hunk 3]
```diff
@@ -9,15 +9,15 @@ namespace GShark.Optimization
...
+        private readonly NurbsBase _curve0;
+        private readonly NurbsBase _curve1;
 
         /// <summary>
         /// Initialize the class, which collects the functions used for the minimization problem.
         /// </summary>
         /// <param name="curve0">First curve used in the intersection process.</param>
         /// <param name="curve1">First curve used in the intersection process.</param>
-        public CurvesIntersectionObjectives(NurbsCurve curve0, NurbsCurve curve1)
+        public CurvesIntersectionObjectives(NurbsBase curve0, NurbsBase curve1)
```
Comment (cesarecaoduro) [inline src/GShark/Optimization/CurvesIntersectionObjectives.cs:20]: Do we need this method to be public?

File: src/GShark/Optimization/CurvePlaneIntersectionObjectives.cs [hunk 4]
```diff
@@ -8,15 +8,15 @@ namespace GShark.Optimization
...
+        private readonly NurbsBase _curve;
         private readonly Plane _plane;
 
         /// <summary>
         /// Initializes the class, which collects the functions used for the minimization problem.
         /// </summary>
         /// <param name="crv">ICurve used in the intersection process.</param>
         /// <param name="plane">Plane used in the intersection process.</param>
-        public CurvePlaneIntersectionObjectives(NurbsCurve crv, Plane plane)
+        public CurvePlaneIntersectionObjectives(NurbsBase crv, Plane plane)
```
Comment (cesarecaoduro) [inline src/GShark/Optimization/CurvePlaneIntersectionObjectives.cs:19]: Do we need this method to be public?

**Review by cesarecaoduro — APPROVED** [review]

File: src/GShark/Geometry/NurbsBase.cs (hunk 2 above)
Comment (sonomirco) [reply src/GShark/Geometry/NurbsBase.cs:587]: There is ``SplitAt`` for a list of parameters.

**Pull request merged**